### ISSUE TITLE
aten::layer_norm perf is extremely slow compared to Caffe2

### ISSUE THREAD

**Issue by jamesr66a** [pytorch/pytorch#11867]
Running on CPU, it appears that `aten::layer_norm`, which is implemented in terms of `thnn_batch_normalization`, is 6.5x slower than the equivalent operator implemented in Caffe2 using eigen. Example benchmark:

```
import torch
import time

input = torch.randn(20, 5, 10, 10, dtype=torch.float)
m = torch.nn.LayerNorm(input.size()[1:])
m.eval()

NITER = 10000

s = time.time()
for i in range(NITER):
   o = m(input)
t_pytorch = time.time() - s
print('pytorch time per iter', t_pytorch / NITER)

from caffe2.python import core, workspace

test_net = core.Net("layer_norm_test")
test_net.LayerNorm(["input"], ["output", "mean", "stddev"], epsilon=1e-5)

import numpy as np
workspace.FeedBlob('input', np.random.rand(20, 5, 10, 10).astype('f'))
workspace.CreateNet(test_net)
s = time.time()
workspace.RunNet(test_net.Name(), NITER)
t_caffe2 = time.time() - s
print('caffe2 time per iter', t_caffe2 / NITER)

print('pytorch / caffe2', t_pytorch / t_caffe2)
```

```
pytorch time per iter 0.00012711503505706786
caffe2 time per iter 1.960470676422119e-05
pytorch / caffe2 6.48390392092241
```

Machine info:
```
!lscpu
```

```
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              56
On-line CPU(s) list: 0-55
Thread(s) per core:  2
Core(s) per socket:  14
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               79
Model name:          Intel(R) Xeon(R) CPU E5-2680 v4 @ 2.40GHz
Stepping:            1
CPU MHz:             2401.000
CPU max MHz:         2401.0000
CPU min MHz:         1200.0000
BogoMIPS:            4788.85
Virtualization:      VT-x
L1d cache:           32K
L1i cache:           32K
L2 cache:            256K
L3 cache:            35840K
NUMA node0 CPU(s):   0-13,28-41
NUMA node1 CPU(s):   14-27,42-55
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 intel_pt tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm rdt_a rdseed adx smap xsaveopt cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts
```

The hot path in the THNN op seems to be this `set1d` call right here: https://github.com/pytorch/pytorch/blob/master/aten/src/THNN/generic/BatchNormalization.c#L56

If we dig into the caffe2 implementation, we find that for GPU there are nontrivial states persisted across invocations of the op: https://github.com/pytorch/pytorch/blob/master/caffe2/operators/layer_norm_op.h#L56. The motivation behind this is that instantiating a caffe2 tensor grabs a global lock, so if you do that in the `Run` method of an operator, you end up serializing your network's execution.

We should investigate what the proper path for op unification is, and see if we can get away with using the caching allocator for these GPU states

**Comment by SsnL** [comment]
Personally I have also found instance norm and group norm to execute the THNN code path (not cudnn).

**Comment by ngimel** [comment]
@jamesr66a As bad as aten::layer_norm forward might be, backward is much worse - did you compare backward with c2 implementation? 
@ssnl the reason why THNN is used is for layer norm the learnable scale and shift are different from bn, so bn kernel is used exclusively for normalization, and scale/shift is applied separately. cudnn likely does not support this mode.

**Comment by ngimel** [comment]
Closing, LayerNorm was rewritten multiple times since, please reopen if issue still persists. cc @BIT-silence
